### ISSUE TITLE
Return all errors that occur in Writer.WriteMessages

### ISSUE THREAD

**Issue by Evanjt1** [segmentio/kafka-go#363]
Currently, the `WriteMessages` receiver only returns a singular error if any message in a given batch fails to deliver to kafka. It's great to know that something went wrong, but it would even better to have information on a per message level.

There are a couple of limitations with the current single error signature:
- The error returned has the potential to be misleading. The single error in some cases is really just the last error read from the `res` channel. The caller ends up losing valuable information about other errors which may have occurred in the batch limiting error handling logic.
- The caller has no choice but to treat that entire batch as a failure.

The second bullet point is especially poignant for me. As an example, If I want to write the failed messages to a dead letter queue, I'm forced to write them all.

Here are a couple different approaches for making a more useful return for the receiver:

1. Use a library such as https://github.com/hashicorp/go-multierror to return all `writerError`s that occurred for the batch. I'm not sold on introducing a dependency, but the solution would be pretty straightforward.

2. Something along the lines of:
```go
func (w *Writer) WriteMessages(ctx context.Context, msgs ...Message) []WriterError

// WriterError is just the current writerError exposed publicly
type WriterError struct {
    msg Message
    err error
} 
```
Although, I'm not a big fan of returning error slices. We could always create a new error type to wrap the WriterErrors with some receivers to allow the caller to dig deeper into the errors if one so chooses. This would also allow us to preserve the signature.

Thoughts?

I'll obligate myself to implementing whatever solution is decided upon if this feature request is deemed useful.

**Comment by achille-roussel** [comment]
Hello @Evanjt1, thanks for opening an issue!

The idea of exposing every error sounds good to me, but we can't make a breaking API change like returning a `[]WriterError` instead of an `error`.

I agree that we probably don't want to take an external dependency for this. The code to implement a multi-error type should be simple, and we can use a type that implements the `error` interface to add the feature without changing the API I think.

For example:
```go
type MessageErrorList []MessageError

func (e MessageErrorList) Error() string { ... }

type MessageError struct {
    Msg Message
    Err error
}
```
Then programs can use type assertion on the error returned by `WriteMessages` if they're interested in getting the detailed information (otherwise they can keep using the returned `error` value).

Let me know what you think about it!

**Comment by Evanjt1** [comment]
@achille-roussel Agreed, I think that's the best way to go about this. Cool if I implement this?

**Comment by Evanjt1** [comment]
@achille-roussel I'm finally revisiting this, anything important I should know before putting in PR?

**Comment by achille-roussel** [comment]
I don't think we made profound changes in this area, so go for it 👍

**Comment by achille-roussel** [comment]
Fixed in v0.4, closing.
